src/bootstrap/src/utils/ra_project.rs

```diff
@@ -0,0 +1,155 @@
+
+use serde_derive::Serialize;
+use std::collections::{BTreeMap, BTreeSet};
+use std::io;
+use std::path::Path;
+
+use crate::core::metadata::{project_metadata, workspace_members, Dependency};
+use crate::Config;
+
+#[derive(Debug, Serialize)]
+/// FIXME(before-merge): doc-comment
+pub(crate) struct RustAnalyzerProject {
+    crates: Vec<Crate>,
+    sysroot: String,
+    sysroot_src: String,
+}
+
+#[derive(Debug, Default, Serialize, PartialEq)]
+struct Crate {
+    cfg: Vec<String>,
+    deps: BTreeSet<Dep>,
+    display_name: String,
+    edition: String,
+    env: BTreeMap<String, String>,
+    is_proc_macro: bool,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    proc_macro_dylib_path: Option<String>,
+    is_workspace_member: bool,
+    root_module: String,
+}
+
+#[derive(Debug, Default, Serialize, PartialEq, PartialOrd, Ord, Eq)]
+struct Dep {
+    #[serde(rename = "crate")]
+    crate_index: usize,
+    name: String,
+}
+
+impl RustAnalyzerProject {
+    #[allow(dead_code)] // FIXME(before-merge): remove this
+    pub(crate) fn collect_ra_project_data(config: &Config) -> Self {
+        let mut ra_project = RustAnalyzerProject {
+            crates: vec![],
+            sysroot: format!("{}", config.out.join("host").join("stage0").display()),
+            sysroot_src: format!("{}", config.src.join("library").display()),
+        };
+
+        let packages: Vec<_> = project_metadata(config).collect();
+        let workspace_members: Vec<_> = workspace_members(config).collect();
+
+        for package in &packages {
+            let is_not_indirect_dependency = packages
+                .iter()
+                .filter(|t| {
+                    let used_from_other_crates = t.dependencies.contains(&Dependency {
+                        name: package.name.clone(),
+                        source: package.source.clone(),
+                    });
+
+                    let is_local = t.source.is_none();
+
+                    (used_from_other_crates && is_local) || package.source.is_none()
+                })
+                .next()
+                .is_some();
+
+            if !is_not_indirect_dependency {
+                continue;
+            }
+
+            for target in &package.targets {
+                let mut krate = Crate::default();
+                krate.display_name = target.name.clone();
+                krate.root_module = target.src_path.clone();
+                krate.edition = target.edition.clone();
+                krate.is_workspace_member = workspace_members.iter().any(|p| p.name == target.name);
+                krate.is_proc_macro = target.crate_types.contains(&"proc-macro".to_string());
+
+                // FIXME(before-merge): We need to figure out how to find proc-macro dylibs.
+                // if krate.is_proc_macro {
+                //     krate.proc_macro_dylib_path =
+                // }
```

Comment:
Any opinion/suggestion on this? @rust-lang/rust-analyzer 

Can't r-a figure this out on its own when this field isn't provided (similar to when it operates without `rust-project.json`)?

Currently, I can access features such as hover (to see more context about a macro attribute), go-to-definition, go-to-reference, etc. However, we get this "proc macro `X` not expanded: No proc-macros present for crate" error as shown below.

![image](https://github.com/rust-lang/rust/assets/39852038/4a5d17c6-df59-4707-9012-6f6173df9199)

Reply:
I'm on vacation until thursday next week so just a quick drive by comment. The way this works with carbgo based projects is that r-a builds proc-macros and runs build scripts in a first step, then extracts the dylib paths from that output. So it can't do that with a rust-project.json. cc @davidbarsky

Reply:
> The way this works with carbgo based projects is that r-a builds proc-macros and runs build scripts in a first step, then extracts the dylib paths from that cargo json output.

Hmm, I think that seems reasonable. I can do the same on bootstrap side, thank you.

Reply:
https://github.com/rust-lang/rust/pull/120611/commits/9219a80fe1b7c91078c07af49752506af0ed470c should do the work. Everything is running smoothly on my end at the moment. I just need a couple of extra days to polish the changes and ensure that I haven't missed anything.